### ISSUE TITLE
Tracy asm time profile suggests slow loads in mlir.dot kernel

### ISSUE THREAD

**Issue by bjacob** [google/iree#3989]
I've let the benchmark run for 10 seconds to get more samples.
The top symbol `serving_default_ex_dispatch_11` got 36,00 samples and the hot loops is show in this screenshot below.
The thing that looks noteworthy here is how the load instructions are reported to be the expensive ones.
If this is significant (as opposed to being a quirk of the CPU's PMU being more prone to interrupting load instructions) then this would suggest cache misses. I'll try to follow up with PMU counter queries...

![dot-asm-bad-loads](https://user-images.githubusercontent.com/79535/100016493-7f180100-2da7-11eb-8dd6-c106267e3edf.jpg)

**Comment by asaadaldien** [comment]
@bjacob Is that profile with pre_alloc branch?, if not loads are expected to take a long time.

**Comment by bjacob** [comment]
D'oh... right, i had rebased and dropped my checkout of your branch. Let me try again with that...

**Comment by bjacob** [comment]
Actually we still have this issue with the `ataei-pre_allocation` branch (rebased on current main).
![still-slow-loads](https://user-images.githubusercontent.com/79535/100149958-ea77d680-2e6c-11eb-8461-eb3e8deef446.jpg)

**Comment by bjacob** [comment]
A little experiment: what happens if we replace all these loads (LDR instructions) by just MOVI instructions zeroing the destination register but not accessing memory?

Tool for this experiment: https://gist.github.com/bjacob/e69850fa22d373fef76c7a842b725975
What's convenient is that since our `.fbvm` files are uncompressed, we can simply run it in place on that.

```
cp /tmp/android_module.fbvm /tmp/android_module_movi.fbvm
hack_arm_binary /tmp/android_module_movi.fbvm ffc00000 3dc00000 ffffffe0 4f00e400
adb push /tmp/android_module_movi.fbvm /data/local/tmp
adb shell taskset 80 /data/local/tmp/iree-benchmark-module --driver=dylib --function_inputs='1x384xi32,1x384xi32,1x384xi32' --module_file=/data/local/tmp/android_module_movi.fbvm --entry_function=serving_default
```

**Caveat:** my build is with Tracy instrumentation, that probably corrupts latencies. However that should affect both sides of the comparison equally.

**Results:** (on MobileBERT on Pixel4, `taskset 80`. I have applied (and rebased) the above-discussed `ataei-pre_allocation` branch )

Before the MOVI tweak: **834 ms**
After the MOVI tweak: **651 ms**

**Comment by benvanik** [comment]
Really good idea @bjacob! That isolates the impacts of the memory access from the impacts on l1 cache utilization/etc. #3817 will make it possible to split the shared objects from the flatbuffer such that the shared object can be easily modified after compilation but if we wanted to do more of these tests we could hack in some ways of doing them with a few changes + a breakpoint you wait on while messing with a file on disk.

For example, here's where we choose to emit an object file but we could add a flag for choosing like clang does with `-S` to instead use `CGFT_AssemblyFile` and write out a .s:
https://github.com/google/iree/blob/e0ec2044f731db2c223809be822df3e361933571/iree%2Fcompiler%2FDialect%2FHAL%2FTarget%2FLLVM%2FLLVMIRPasses.cpp#L109-L113

Then here's where we run the linker command taking the .o -> .so: https://github.com/google/iree/blob/e0ec2044f731db2c223809be822df3e361933571/iree%2Fcompiler%2FDialect%2FHAL%2FTarget%2FLLVM%2FAOT%2Finternal%2FUnixLinkerTool.cpp#L73

Instead, you could do `$CC temp.s -shared -o temp.so` to assemble that .s and continue the embedding.

This would let you edit the .s file by hand to do whatever you wanted without needing to edit the executable bytes by hand -- as badass as that is :)

**Comment by benvanik** [comment]
Riffing here, but related if we want to file issues to track such things: the other thing we can do to make that something everyone can use (vs hacking the bytes/asm) is add some MLIR passes on the LLVM IR dialect (or maybe vector above it) and use that to make such transforms ("neuter all reads and/or writes" or "reduce all read bits from 32 to 8 to simulate memory access on in-memory quantized buffers" etc. One of the coolest features of pix is dr. pix: https://devblogs.microsoft.com/pix/gpu-captures/#dr-pix which lets you play "what if" - in the example linked there it benchmarks your original commands and then the commands where the entire viewport is turned into 1x1 so you can isolate the impact of the dataflow/vertex pipeline independent from the pixel processing pipeline (as there's only 1 pixel now) - but otherwise all commands are the same. With such pass flags in place for doing these transformations we could do the same thing with a shell script: compile with/without the flags, run iree-benchmark-module, diff the results. There are counters and such we can use to infer a lot of similar information on some systems but not all systems have ones that are easily indicative and it's harder to get a layperson to internalize what a slightly higher l1 cache writeback count implies and know what they should change in their model to make things faster whereas such a tool could guide them along with "reduce the amount of data you are reading for a max potential N% performance improvement." I dub it Dr. IREE 👩‍⚕️ :P

**Comment by asaadaldien** [comment]
> Actually we still have this issue with the `ataei-pre_allocation` branch (rebased on current main).
> ![still-slow-loads](https://user-images.githubusercontent.com/79535/100149958-ea77d680-2e6c-11eb-8461-eb3e8deef446.jpg)

10% in loading the second row seems like a cache miss (I need to add PMU counters soon =) ). One thing we can try here is to experiment with promoting the tile to stack memory so loads aren't jumping with big strides.

**Comment by bjacob** [comment]
Thanks Ben for the pointers!
Regarding making that a nice reusable feature in the compiler... Ideally, I'd like a way to annotate whatever the "source" code is, e.g. MLIR vector dialect, and have that information propagated all the way down. Discussed that with @nicolasvasilache . Sounded like it was going to be a big project to make that consistently propagated all the way through LLVM compilation.

**Comment by asaadaldien** [comment]
> Really good idea @bjacob! That isolates the impacts of the memory access from the impacts on l1 cache utilization/etc. #3817 will make it possible to split the shared objects from the flatbuffer such that the shared object can be easily modified after compilation but if we wanted to do more of these tests we could hack in some ways of doing them with a few changes + a breakpoint you wait on while messing with a file on disk.
> 
> For example, here's where we choose to emit an object file but we could add a flag for choosing like clang does with `-S` to instead use `CGFT_AssemblyFile` and write out a .s:
> https://github.com/google/iree/blob/e0ec2044f731db2c223809be822df3e361933571/iree%2Fcompiler%2FDialect%2FHAL%2FTarget%2FLLVM%2FLLVMIRPasses.cpp#L109-L113
> 
> Then here's where we run the linker command taking the .o -> .so:
> 
> https://github.com/google/iree/blob/e0ec2044f731db2c223809be822df3e361933571/iree%2Fcompiler%2FDialect%2FHAL%2FTarget%2FLLVM%2FAOT%2Finternal%2FUnixLinkerTool.cpp#L73
> 
> Instead, you could do `$CC temp.s -shared -o temp.so` to assemble that .s and continue the embedding.
> 
> This would let you edit the .s file by hand to do whatever you wanted without needing to edit the executable bytes by hand -- as badass as that is :)

Nice idea! we can optionally embed the textual assembly instead of executable bytes. And have another tool that does replace assembly and generate embedded dylib `e.g iree-as`

**Comment by nicolasvasilache** [comment]
Right, so let me give some details here because there may always be shortcuts depending on the concrete use case.

There are 3 big things that I see off the bat to help achieve "I'd like a way to annotate whatever the "source" code is, e.g. MLIR vector dialect":
1. the source may be "very very far" from the vector abstraction you want to tweak: if want to do this as part of the IREE compilation flow then controlling at the very top without getting in the guts of the compiler will be hard. There may be an opportunity to use location information to hackily inject something the gets propagated but that relies on everything above, until the moment the vector ops come into existence, to propagate everything properly.
2. the assembly is "very very far" from the modification you want to make. So it has to belong to a good class of "type erased" modifications such that neither MLIR, LLVM or LLC will turn your computation into a pumpkin. Something that type erases the strides and lets you modify them in an external C function may be the right level of opacity but it will still end up inducing modified assembly in various ways that may be quite surprising.
3. even if 1 and 2 existed, providing the magic string that a non-compiler writer would have to type to make the modification you want is probably daunting: think of it as designing the control logic in a processor: SIMD, ALU, memory all that stuff is easy, control logic? that's hard.

To try and be useful still I suggested either of the following:
1. at the op level in the IR, isolate the function you want, copy-paste the assembly and hook it up to a rewrite using the inline_asm stuff I am in the process of adding. 
2. print the IR after all all grab the level when the proper vector abstractions have come into existence. At that level inject a MLIR that does the suggestion above: "Something that type erases the strides and lets you modify them in an external C function".

None of these may be better than just continue along the direction you have been going so far, so YMMV

**Comment by MaheshRavishankar** [comment]
AFAIK, this isnt an issue anymore. This was a missed fusion where fusion wasnt happening at register level. Now IREE CPU backend does do fusion at this level.

@bjacob please close if you think that this issue is resolved

**Comment by bjacob** [comment]
Basically, this issue was a key piece of the evidence motivating the packing approach as in mmt4d.
